app/lib/shared/task_scheduler.dart

```diff
@@ -41,16 +36,34 @@ class TaskScheduler {
   TaskScheduler(this.taskRunner, this.sources);
 
   Future run() async {
-    final Stream<Task> stream =
-        StreamGroup.merge(sources.map((source) => source.startStreaming()));
+    final StreamController newTaskTrigger = new StreamController();
+
+    final List<Queue<Task>> priorityQueues =
+        new List.generate(sources.length, (_) => new Queue());
+
+    // Listen on the streams and put tasks into their own queues.
+    for (int i = 0; i < sources.length; i++) {
+      final TaskSource source = sources[i];
+      final Queue<Task> queue = priorityQueues[i];
+      source.startStreaming().listen((Task task) {
+        queue.add(task);
+        newTaskTrigger.add(null);
+      });
+    }
 
-    await for (Task task in stream) {
+    // On new task, pick the first queue that has a task and execute that.
+    await for (var _ in newTaskTrigger.stream) {
+      final Queue<Task> queue = priorityQueues.firstWhere((q) => q.isNotEmpty);
```

Comment:
there is 1:1 mapping between adding one item to any queue + triggering the newTaskTrigger.stream. Also, on each item of that stream we remove exactly one task. At this point one of the queues should be always non-empty.